internal/services/serviceprincipals/synchronization_job_resource.go

```diff
@@ -0,0 +1,251 @@
+	timeout, _ := ctx.Deadline()
+	_, err = (&resource.StateChangeConf{
+		Pending:                   []string{"Waiting"},
+		Target:                    []string{"Done"},
+		Timeout:                   time.Until(timeout),
+		MinTimeout:                1 * time.Second,
+		ContinuousTargetOccurence: 5,
+		Refresh: func() (interface{}, string, error) {
+			_, status, err := client.Get(ctx, id.JobId, id.ServicePrincipalId)
+			if err != nil {
+				if status == http.StatusNotFound {
+					return "stub", "Waiting", nil
+				}
+				return nil, "Error", fmt.Errorf("retrieving member")
+			}
+			return "stub", "Done", nil
+		},
+	}).WaitForStateContext(ctx)
+
+	if err != nil {
+		return tf.ErrorDiagF(err, "Waiting for synchronization job %q", id.JobId)
+	}
+
+	d.SetId(id.String())
+
+	// Start job if desired
+	desiredState := d.Get("state").(string)
+	if desiredState == "Active" {
+		_, err := client.Start(ctx, id.JobId, id.ServicePrincipalId)
+		if err != nil {
+			return tf.ErrorDiagF(err, "Starting synchronization job %q", id.JobId)
+		}
+	}
+
+	return synchronizationJobResourceRead(ctx, d, meta)
+}
+
+func synchronizationJobResourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.SynchronizationJobClient
+
+	id, err := parse.SynchronizationJobID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing synchronization job with ID %q", d.Id())
+	}
+
+	job, status, err := client.Get(ctx, id.JobId, id.ServicePrincipalId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			log.Printf("[DEBUG] Synchronization job with ID %q for service principal %q was not found - removing from state!", id.JobId, id.ServicePrincipalId)
+			d.SetId("")
+			return nil
+		}
+		return tf.ErrorDiagF(err, "Retrieving synchronization job with object ID %q", id.JobId)
+	}
+	tf.Set(d, "id", job.ID)
+	tf.Set(d, "schedule", flattenSynchronizationSchedule(job.Schedule))
+	tf.Set(d, "template_id", job.TemplateId)
+	tf.Set(d, "state", job.Schedule.State)
+	return nil
+}
+
+func synchronizationJobResourceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.SynchronizationJobClient
+	id, err := parse.SynchronizationJobID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing synchronization job with ID %q", d.Id())
+	}
+	desiredState := d.Get("state").(string)
+	if desiredState != d.Get("schedule.state") {
+		if desiredState == "Active" {
+			_, err := client.Start(ctx, id.JobId, id.ServicePrincipalId)
+			if err != nil {
+				return tf.ErrorDiagF(err, "Starting synchronization job %q", id.JobId)
+			}
+		}
+		if desiredState == "Disabled" {
+			_, err := client.Pause(ctx, id.JobId, id.ServicePrincipalId)
+			if err != nil {
+				return tf.ErrorDiagF(err, "Pausing synchronization job %q", id.JobId)
+			}
+		}
+	}
```

Comment:
Can this be reworked as per a boolean property `enabled`? Also, whilst it's legal to call `d.Get()` with nested properties, we typically prefer to roll container properties (blocks, lists, sets etc) into flatten/expand functions.

Reply:
I resolved this by using `d.HasChange("enabled")` since the state is already set to "enabled" in the read call.